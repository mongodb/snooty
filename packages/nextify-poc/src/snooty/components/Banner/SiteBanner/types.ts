export interface SiteBannerContent {
  isEnabled: boolean;
  altText: string;
  imgPath?: string;
  tabletImgPath?: string;
  mobileImgPath?: string;
  bgColor?: string;
  text?: string;
  pillText?: string;
  url: string;
}
