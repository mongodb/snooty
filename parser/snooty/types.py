import enum
import hashlib
from dataclasses import dataclass
from pathlib import PurePath
from typing import Any, Callable, Dict, Set, List, Tuple, Optional, Union

SerializableType = Union[None, bool, str, int, float, Dict[str, Any], List[Any]]
EmbeddedRstParser = Callable[[str, int, bool], List[SerializableType]]


@dataclass
class Diagnostic:
    __slots__ = ('message', 'severity', 'start', 'end')

    class Level(enum.IntEnum):
        info = 0
        error = 1
        warning = 2

        @classmethod
        def from_docutils(cls, docutils_level: int) -> 'Diagnostic.Level':
            level = docutils_level - 1
            level = min(level, cls.warning)
            level = max(level, cls.info)
            return cls(level)

    severity: Level
    message: str
    start: Tuple[int, int]
    end: Tuple[int, int]

    @property
    def severity_string(self) -> str:
        return self.severity.name.title()

    @classmethod
    def create(cls, severity: Level, message: str,
               start: Union[int, Tuple[int, int]],
               end: Union[None, int, Tuple[int, int]] = None) -> 'Diagnostic':
        if isinstance(start, int):
            start_line, start_column = start, 0
        else:
            start_line, start_column = start

        if end is None:
            end_line, end_column = start_line, 1000
        elif isinstance(end, int):
            end_line, end_column = end, 1000
        else:
            end_line, end_column = end

        return cls(severity, message, (start_line, start_column), (end_line, end_column))

    @classmethod
    def warning(cls, message: str,
                start: Union[int, Tuple[int, int]],
                end: Union[None, int, Tuple[int, int]] = None) -> 'Diagnostic':
        return cls.create(cls.Level.warning, message, start, end)

    @classmethod
    def error(cls, message: str,
              start: Union[int, Tuple[int, int]],
              end: Union[None, int, Tuple[int, int]] = None) -> 'Diagnostic':
        return cls.create(cls.Level.error, message, start, end)


@dataclass
class StaticAsset:
    __slots__ = ('fileid', 'checksum', 'data')

    fileid: str
    checksum: str
    data: Optional[bytes]

    def __hash__(self) -> int:
        return hash(self.checksum)

    @classmethod
    def load(cls, fileid: str, path: PurePath) -> 'StaticAsset':
        with open(path, 'rb') as f:
            data = f.read()
        asset_hash = hashlib.blake2b(data, digest_size=32).hexdigest()
        return cls(fileid, asset_hash, data)


@dataclass
class Page:
    source_path: PurePath
    source: str
    ast: SerializableType
    diagnostics: List[Diagnostic]
    static_assets: Set[StaticAsset]
    category: Optional[str] = None

    def get_id(self) -> PurePath:
        if self.category:
            # Giza wrote out yaml file artifacts under a directory. e.g. steps-foo.yaml becomes
            # steps/foo.rst
            return self.source_path.parent.joinpath(
                PurePath(self.category),
                self.source_path.name.replace(f'{self.category}-', '', 1))
        return self.source_path
