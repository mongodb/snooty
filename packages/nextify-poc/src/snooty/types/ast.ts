// Need to figure out which is node type and which is component type
type ComponentType =
  | Exclude<NodeType, 'block-quote' | 'directive' | 'directive_argument' | 'role' | 'target_identifier'>
  | 'banner'
  | 'blockquote'
  | 'button'
  | 'card'
  | 'card-group'
  | 'chapter'
  | 'chapters'
  | 'code'
  | 'collapsible'
  | 'community-driver'
  | 'io-code-block'
  | 'cond'
  | 'container'
  | 'cta'
  | 'cta-banner'
  | 'definitionList'
  | 'definitionListItem'
  | 'deprecated'
  | 'deprecated-version-selector'
  | 'describe'
  | 'emphasis'
  | 'extract'
  | 'field'
  | 'field_list'
  | 'figure'
  | 'footnote'
  | 'footnote_reference'
  | 'glossary'
  | 'guide-next'
  | 'heading'
  | 'hlist'
  | 'image'
  | 'include'
  | 'introduction'
  | 'kicker'
  | 'line'
  | 'line_block'
  | 'literal_block';

type NodeType =
  | 'admonition'
  | 'banner'
  | 'block-quote'
  | 'code'
  | 'root'
  | 'section'
  | 'heading'
  | 'reference'
  | 'directive'
  | 'directive_argument'
  | 'line'
  | 'line_block'
  | 'list'
  | 'list-table'
  | 'listItem'
  | 'text'
  | 'literal'
  | 'literalinclude'
  | 'definitionList'
  | 'definitionListItem'
  | 'emphasis'
  | 'method-selector'
  | 'only'
  | 'openapi-changelog'
  | 'paragraph'
  | 'procedure'
  | 'reference'
  | 'ref_role'
  | 'role'
  | 'release_specification'
  | 'rubric'
  | 'search-results'
  | 'section'
  | 'seealso'
  | 'sharedinclude'
  | 'strong'
  | 'substitution_reference'
  | 'tabs-selector'
  | 'time'
  | 'title_reference'
  | 'transition'
  | 'versionadded'
  | 'versionchanged'
  | 'tabs'
  | 'target'
  | 'target_identifier'
  | 'text'
  | 'wayfinding';

type RoleName =
  | 'abbr'
  | 'class'
  | 'command'
  | 'file'
  | 'guilabel'
  | 'icon'
  | 'highlight-blue'
  | 'highlight-green'
  | 'highlight-red'
  | 'highlight-yellow'
  | 'icon-fa5'
  | 'icon-fa5-brands'
  | 'icon-fa4'
  | 'icon-mms'
  | 'icon-charts'
  | 'icon-lg'
  | 'kbd'
  | 'red'
  | 'gold'
  | 'required'
  | 'sub'
  | 'subscript'
  | 'sup'
  | 'superscript'
  | 'link-new-tab';

type NodeName =
  | RoleName
  | AdmonitionName
  | 'list-table'
  | 'contents'
  | 'collapsible'
  | 'tabs'
  | 'tab'
  | 'facet'
  | 'toctree'
  | 'deprecated'
  | 'versionadded'
  | 'versionchanged';

interface Node {
  type: NodeType;
}

interface ParentNode extends Node {
  children: Node[];
}

interface Root extends ParentNode {
  options: Record<string, any>;
  fileid: string;
}

interface BlockQuoteNode extends ParentNode {
  type: 'block-quote';
}

interface HeadingNode extends ParentNode {
  type: 'heading';
  id: string;
}

interface ParagraphNode extends ParentNode {
  type: 'paragraph';
}

interface EmphasisNode extends ParentNode {
  type: 'emphasis';
}

interface StrongNode extends ParentNode {
  type: 'strong';
}

interface ReferenceNode extends ParentNode {
  type: 'reference';
  refuri: string;
}

interface Directive<TOptions = { [key: string]: string }> extends ParentNode {
  type: 'directive';
  name: NodeName;
  argument: Node[];
  domain?: string;
  options?: TOptions;
}

interface ListTableNode extends Directive {
  name: 'list-table';
  children: ListNode[];
  options?: {
    widths?: string;
    'header-rows'?: string;
  };
}

interface ListNode extends ParentNode {
  type: 'list';
  enumtype: 'unordered' | 'ordered';
  children: ListItemNode[];
}

interface ListItemNode extends ParentNode {
  type: 'listItem';
}

interface LiteralNode extends ParentNode {
  type: 'literal';
}

interface LineBlockNode extends ParentNode {
  type: 'line_block';
}

interface LineNode extends ParentNode {
  type: 'line';
}

interface TextNode extends Node {
  type: 'text';
  value: string;
}

interface DefinitionListNode extends ParentNode {
  type: 'definitionList';
}

interface DefinitionListItemNode extends ParentNode {
  type: 'definitionListItem';
  term: Node[];
}

interface CodeNode extends Node {
  type: 'code';
  lang: string;
  copyable: boolean;
  emphasize_lines: number[];
  value: string;
  linenos: boolean;
}

interface MethodNode extends ParentNode {
  type: 'target';
  name: 'method';
  domain: 'mongodb';
}

interface DirectiveArgumentNode extends ParentNode {
  type: 'directive_argument';
}

interface TargetIdentifierNode extends ParentNode {
  type: 'target_identifier';
  ids: string[];
}

interface valueNode extends Node {
  value: string;
}
interface titleReferenceNode {
  children: valueNode[];
}

type TwitterOptions = {
  creator: string;
  image: string;
  'image-alt': string;
  site: string;
  title: string;
};

interface TwitterNode extends Directive<TwitterOptions> {
  options: TwitterOptions;
}
interface VersionModifiedNode extends Directive {
  children: Node[];
}

type CollapsibleOptions = {
  heading?: string;
  sub_heading?: string;
  id?: string;
};

interface CollapsibleNode extends Directive<CollapsibleOptions> {
  type: 'directive';
  name: 'collapsible';
  options?: CollapsibleOptions;
}

interface ContentsOptions {
  local?: boolean;
  backlinks?: string;
  depth?: number;
  class?: string;
}

interface ContentsNode extends Directive<ContentsOptions> {
  type: 'directive';
  name: 'contents';
  options: ContentsOptions;
}

interface TabsNode extends Directive {
  type: 'directive';
  name: 'tabs';
}

interface TabOptions {
  tabid: string;
}

interface TabNode extends Directive<TabOptions> {
  type: 'directive';
  name: 'tab';
  options?: TabOptions;
}

interface FacetOptions {
  name: string;
  values: string;
}

interface FacetNode extends Directive<FacetOptions> {
  type: 'directive';
  name: 'facet';
  options?: FacetOptions;
}

type AdmonitionName = 'example' | 'note' | 'tip' | 'see' | 'seealso' | 'warning' | 'important';

interface AdmonitionNode extends Directive {
  type: 'directive';
  name: AdmonitionName;
}

interface TocTreeEntry {
  title?: string;
  slug?: string;
}

interface TocTreeOptions {
  osiris_parent: boolean;
}

interface TocTreeDirective extends Directive<TocTreeOptions> {
  type: 'directive';
  name: 'toctree';
  entries: TocTreeEntry[];
}

export type {
  ComponentType,
  NodeType,
  RoleName,
  NodeName,
  Node,
  ParentNode,
  Root,
  HeadingNode,
  ReferenceNode,
  TextNode,
  CodeNode,
  Directive,
  ListNode,
  ListTableNode,
  ListItemNode,
  ParagraphNode,
  LiteralNode,
  LineBlockNode,
  LineNode,
  DefinitionListNode,
  DefinitionListItemNode,
  MethodNode,
  TargetIdentifierNode,
  DirectiveArgumentNode,
  CollapsibleNode,
  CollapsibleOptions,
  ContentsNode,
  EmphasisNode,
  StrongNode,
  TabsNode,
  TabNode,
  FacetNode,
  AdmonitionNode,
  AdmonitionName,
  TocTreeEntry,
  TocTreeDirective,
  TwitterOptions,
  TwitterNode,
  VersionModifiedNode,
  titleReferenceNode,
  BlockQuoteNode,
};
