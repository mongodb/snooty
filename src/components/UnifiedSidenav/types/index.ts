export interface BreadCrumb {
  path: string;
  title: string;
}

export interface TocItem {
  label: string;
  url?: string;
  group?: boolean;
  contentSite: string;
  collapsible?: boolean;
  breadcrumbs?: BreadCrumb[];
  showSubNav?: boolean;
  isExternal?: boolean;
  versionDropdown?: boolean;
  items?: TocItem[];
}

export interface ActiveVersions {
  [project: string]: string;
}
