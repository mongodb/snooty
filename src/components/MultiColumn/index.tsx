import React from 'react';
import { css, cx } from '@leafygreen-ui/emotion';
import { ParentNode } from '../../types/ast';
import ComponentFactory from '../ComponentFactory';
import { theme } from '../../theme/docsTheme';

const MultiColumn = ({ nodeData }: { nodeData: ParentNode }) => {
  console.log('nodeData', nodeData);
  return (
    <div
      className={cx(css`
        display: flex;
        gap: ${theme.size.xxlarge};
        margin-top: 60px;
        margin-bottom: 40px;

        @media ${theme.screenSize.upToLarge} {
          flex-direction: column;
          gap: ${theme.size.xlarge};
        }

        @media ${theme.screenSize.upToSmall} {
          gap: 40px;
        }
      `)}
    >
      {nodeData.children.map((child, i) => {
        return <ComponentFactory nodeData={child} key={i} />;
      })}
    </div>
  );
};

export default MultiColumn;
