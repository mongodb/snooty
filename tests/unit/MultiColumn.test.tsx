import React from 'react';
import { render } from '@testing-library/react';
import MultiColumn from '../../src/components/MultiColumn';
// data for MultiComponent
import { ParentNode } from '../../src/types/ast';
import mockData from './data/MultiColumn.test.json';

const renderMultiColumn = () => render(<MultiColumn nodeData={mockData as ParentNode} />);

describe('Rendering a MultiComponent', () => {
  it('renders correct', () => {
    const wrapper = renderMultiColumn();
    expect(wrapper.asFragment()).toMatchSnapshot();
  });
});
