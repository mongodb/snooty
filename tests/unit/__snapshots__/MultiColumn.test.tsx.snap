// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering a MultiComponent renders correct 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 128px;
  margin-top: 60px;
  margin-bottom: 40px;
}

@media only screen and (max-width: 1024px) {
  .emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    gap: 64px;
  }
}

@media only screen and (max-width: 480px) {
  .emotion-0 {
    gap: 40px;
  }
}

.emotion-1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-1 .intro-code-block {
  margin-top: 0;
}

.emotion-1 ul {
  list-style: none;
  padding-left: 30px;
  margin: 0;
  position: relative;
}

.emotion-1 ul li::before {
  content: url('data:image/svg+xml,<svg xmlns="http://www.w3.org/2000/svg" width="10" height="10" viewBox="0 0 10 10" fill="none"><circle cx="5" cy="5" r="4.25" stroke="%2300ED64" stroke-width="1.9"/></svg>');
  position: absolute;
  left: -1px;
  -webkit-transform: translateY(10%);
  -moz-transform: translateY(10%);
  -ms-transform: translateY(10%);
  transform: translateY(10%);
}

.emotion-2 {
  display: table;
  margin: 24px 0;
  min-width: 150px;
  table-layout: fixed;
  width: 100%;
}

.emotion-2 button>div>div {
  font-size: 13px;
}

.emotion-2 pre {
  background-color: #F9FBFA;
  color: #001E2B;
}

.dark-theme .emotion-2 pre {
  background-color: #001E2B;
  color: #F9FBFA;
}

.emotion-2 [data-testid='leafygreen-code-panel'] {
  background-color: #FFFFFF;
  border-color: #E8EDEB;
}

.dark-theme .emotion-2 [data-testid='leafygreen-code-panel'] {
  background-color: #3D4F58;
  border-color: #3D4F58;
}

.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-keyword,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-keyword.lg-highlight-function,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-keyword.lg-highlight-class,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-tag,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-attr,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-pseudo,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-id,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-class {
  color: #CC3887;
}

.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-keyword,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-keyword.lg-highlight-function,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-keyword.lg-highlight-class,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-tag,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-attr,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-pseudo,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-id,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-selector-class {
  color: #FF7DC3;
}

.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-regexp,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-number,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-literal,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-function.lg-highlight-title {
  color: #016ee9;
}

.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-regexp,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-number,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-literal,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-function.lg-highlight-title {
  color: #2dc4ff;
}

.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-quote,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-section,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-name {
  color: #007ab8;
}

.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-quote,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-section,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-name {
  color: #a5e3ff;
}

.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-string,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-addition {
  color: #12824D;
}

.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-string,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-addition {
  color: #35DE7B;
}

.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-meta,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-meta-string {
  color: #956d00;
}

.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-meta,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-meta-string {
  color: #EDB210;
}

.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-variable,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-deletion,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-symbol,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-bullet,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-meta,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-link,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-attr,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-attribute,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-language,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-template-variable,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-built_in,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-type,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-params {
  color: #D83713;
}

.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-variable,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-deletion,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-symbol,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-bullet,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-meta,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-link,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-attr,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-attribute,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-language,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-template-variable,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-built_in,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-type,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-params {
  color: #FF6F44;
}

.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-title,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-class.lg-highlight-title {
  color: #001E2B;
}

.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-title,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-class.lg-highlight-title {
  color: #F9FBFA;
}

.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-doctag,
.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-formula {
  color: #001E2B;
}

.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-doctag,
.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-formula {
  color: #F9FBFA;
}

.emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-comment {
  color: #3D4F58;
}

.dark-theme .emotion-2 [class*='lg-highlight-hljs-'] .lg-highlight-comment {
  color: #C1C7C6;
}

.emotion-3>div>div {
  display: grid;
  grid-template-columns: code panel;
}

.emotion-3>div {
  border-top-left-radius: 12px;
  border-top-right-radius: 12px;
  border-color: #E8EDEB;
}

.dark-theme .emotion-3>div {
  border-color: #3D4F58;
}

.emotion-4 {
  border: 1px solid #E8EDEB;
  border-radius: 12px;
  overflow: hidden;
}

.emotion-5 {
  position: relative;
  display: grid;
  grid-template-areas: 'code panel';
  grid-template-columns: auto 38px;
  border-radius: inherit;
  z-index: 0;
}

.emotion-5:before,
.emotion-5:after {
  content: '';
  display: block;
  position: absolute;
  z-index: 1;
  top: 0;
  height: 100%;
  width: 40px;
  border-radius: 40%;
  box-shadow: unset;
  -webkit-transition: box-shadow 100ms ease-in-out;
  transition: box-shadow 100ms ease-in-out;
}

.emotion-5:before {
  grid-column: 1;
  left: -40px;
}

.emotion-5:after {
  grid-column: 2;
}

.emotion-6 {
  grid-area: code;
  overflow-x: auto;
  border-radius: inherit;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border: 0;
  padding-top: 8px;
  padding-bottom: 8px;
  margin: 0;
  position: relative;
  -webkit-transition: box-shadow 100ms ease-in-out;
  transition: box-shadow 100ms ease-in-out;
  white-space: pre;
  background-color: #F9FBFA;
  color: #001E2B;
}

@media only screen and (max-device-width: 812px) and (-webkit-min-device-pixel-ratio: 2) {
  .emotion-6 {
    white-space: pre-wrap;
  }
}

@media only screen and (min-device-width: 813px) and (-webkit-min-device-pixel-ratio: 2) {
  .emotion-6 {
    white-space: pre;
  }
}

.emotion-6:focus-visible {
  outline: none;
  box-shadow: 0 0 0 2px #0498EC inset;
}

.emotion-7 {
  color: inherit;
  font-family: 'Source Code Pro',Menlo,monospace;
  font-size: 13px;
  line-height: 20px;
}

.emotion-8 {
  border-spacing: 0;
}

.emotion-9 {
  border-spacing: 0;
  vertical-align: top;
  padding: 0 16px;
}

.emotion-20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  gap: 4px;
  background-color: #FFFFFF;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 6px;
  border-left: solid 1px;
  border-color: #E8EDEB;
  z-index: 2;
  grid-area: panel;
}

.emotion-20 svg {
  width: 16px;
  height: 16px;
}

.emotion-21 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 13px;
  line-height: 20px;
  color: #001E2B;
  font-weight: 400;
  margin-bottom: 16px;
  color: var(--font-color-primary);
}

.emotion-21 strong,
.emotion-21 b {
  font-weight: 700;
}

.emotion-26 {
  grid-area: code;
  overflow-x: auto;
  border-radius: inherit;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border: 0;
  padding-top: 8px;
  padding-bottom: 8px;
  margin: 0;
  position: relative;
  -webkit-transition: box-shadow 100ms ease-in-out;
  transition: box-shadow 100ms ease-in-out;
  white-space: pre;
  background-color: #F9FBFA;
  color: #001E2B;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-top: 6px;
  padding-bottom: 6px;
}

@media only screen and (max-device-width: 812px) and (-webkit-min-device-pixel-ratio: 2) {
  .emotion-26 {
    white-space: pre-wrap;
  }
}

@media only screen and (min-device-width: 813px) and (-webkit-min-device-pixel-ratio: 2) {
  .emotion-26 {
    white-space: pre;
  }
}

.emotion-26:focus-visible {
  outline: none;
  box-shadow: 0 0 0 2px #0498EC inset;
}

.emotion-32 {
  margin: unset;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #001E2B;
  font-size: 24px;
  line-height: 32px;
  font-weight: 500;
  color: #001E2B;
  color: #fff;
  margin-bottom: 37px;
  margin-top: -5px;
}

.emotion-33 {
  margin-top: 0px;
}

.emotion-34::marker {
  color: var(--font-color-primary);
}

.emotion-34>p {
  margin-bottom: 8px;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <script
        class="structured_data"
        type="application/ld+json"
      >
        {"@context":"https://schema.org","@type":"SoftwareSourceCode","codeSampleType":"code snippet","text":"{\\n   first name: \\"Bob\\"\\n   lastname: \\"Smith\\"\\n   email: \\"bob@smith.com\\"\\n   address: {\\n      street: \\"100 Main St\\"\\n      city: \\"Anytown\\"\\n      state: \\"MO\\"\\n      zip: \\"11111\\"\\n   }\\n}","programmingLanguage":"JavaScript"}
      </script>
      <div
        class=" emotion-2 emotion-3 intro-code-block"
      >
        <div
          class="emotion-4"
        >
          <div
            class="emotion-5"
          >
            <pre
              class="emotion-6"
              tabindex="-1"
            >
              <code
                class="lg-highlight-hljs-light javascript emotion-7"
              >
                <table
                  class="emotion-8"
                >
                  <tbody>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                        {
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                           first 
                        <span
                          class="lg-highlight-attr"
                        >
                          name
                        </span>
                        : 
                        <span
                          class="lg-highlight-string"
                        >
                          "Bob"
                        </span>
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                           
                        <span
                          class="lg-highlight-attr"
                        >
                          lastname
                        </span>
                        : 
                        <span
                          class="lg-highlight-string"
                        >
                          "Smith"
                        </span>
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                           
                        <span
                          class="lg-highlight-attr"
                        >
                          email
                        </span>
                        : 
                        <span
                          class="lg-highlight-string"
                        >
                          "bob@smith.com"
                        </span>
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                           
                        <span
                          class="lg-highlight-attr"
                        >
                          address
                        </span>
                        : {
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                              
                        <span
                          class="lg-highlight-attr"
                        >
                          street
                        </span>
                        : 
                        <span
                          class="lg-highlight-string"
                        >
                          "100 Main St"
                        </span>
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                              
                        <span
                          class="lg-highlight-attr"
                        >
                          city
                        </span>
                        : 
                        <span
                          class="lg-highlight-string"
                        >
                          "Anytown"
                        </span>
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                              
                        <span
                          class="lg-highlight-attr"
                        >
                          state
                        </span>
                        : 
                        <span
                          class="lg-highlight-string"
                        >
                          "MO"
                        </span>
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                              
                        <span
                          class="lg-highlight-attr"
                        >
                          zip
                        </span>
                        : 
                        <span
                          class="lg-highlight-string"
                        >
                          "11111"
                        </span>
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                           }
                      </td>
                    </tr>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                        }
                      </td>
                    </tr>
                  </tbody>
                </table>
              </code>
            </pre>
            <div
              class="emotion-20"
              data-testid="leafygreen-code-panel"
            />
          </div>
        </div>
      </div>
      <p
        class="emotion-21"
      >
        This is a paragraph between the code blocks.
      </p>
      <script
        class="structured_data"
        type="application/ld+json"
      >
        {"@context":"https://schema.org","@type":"SoftwareSourceCode","codeSampleType":"code snippet","text":"users.find({ address.zip : \\"11111\\"})","programmingLanguage":"JavaScript"}
      </script>
      <div
        class=" emotion-2 emotion-3 intro-code-block"
      >
        <div
          class="emotion-4"
        >
          <div
            class="emotion-5"
          >
            <pre
              class="emotion-26"
              tabindex="-1"
            >
              <code
                class="lg-highlight-hljs-light javascript emotion-7"
              >
                <table
                  class="emotion-8"
                >
                  <tbody>
                    <tr
                      class=""
                    >
                      <td
                        class="emotion-9"
                      >
                        users.
                        <span
                          class="lg-highlight-title lg-highlight-function"
                        >
                          find
                        </span>
                        ({ address.
                        <span
                          class="lg-highlight-property"
                        >
                          zip
                        </span>
                         : 
                        <span
                          class="lg-highlight-string"
                        >
                          "11111"
                        </span>
                        })
                      </td>
                    </tr>
                  </tbody>
                </table>
              </code>
            </pre>
            <div
              class="emotion-20"
              data-testid="leafygreen-code-panel"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="emotion-1"
    >
      <h3
        class="emotion-32"
      >
        MongoDB also offers
      </h3>
      <ul
        class="emotion-33"
      >
        <li
          class="emotion-34"
        >
          <p
            class="emotion-21"
          >
            Strong consistency with ACID transactions.
          </p>
        </li>
        <li
          class="emotion-34"
        >
          <p
            class="emotion-21"
          >
            Modern additional built-in query capabilities such as geo search, lexical search, and vector search.
          </p>
        </li>
        <li
          class="emotion-34"
        >
          <p
            class="emotion-21"
          >
            Serverless horizontal scaling with geo-aware fault tolerance across all major clouds.
          </p>
        </li>
        <li
          class="emotion-34"
        >
          <p
            class="emotion-21"
          >
            Security primitives that allow MongoDB to operate in the most demanding of enterprise environments.
          </p>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
